### PR TITLE
Attach deadlines questions

### PR DESCRIPTION
**Pull request by AthinaDavari**
This PR closes/references issue #1069 ("Allow to attach deadlines to questions", from discussions). 

In order to close the issue:
- We made two fields in questions' model (and its form) for the change of the feature:
  - ```question_required```, a charfield
  - ```deadline```, a datetime field
- We removed the ```required``` field, since all cases are covered with ```question_required```.
- Also we modified ```src/pretalx/common/mixins/forms.py``` so that the question is required and disabled based on the ```question_required``` field and the current date compared to the ```deadline```. (For "freeze after" questions, of which the deadline has passed, we made the field disable.)
- We changed ```src/pretalx/orga/templates/orga/cfp/question_view.html```. Now questions' table shows require based on the new  ```question_required``` and ```deadline``` fields.
- We changed ```src/pretalx/orga/forms/cfp.py``` . The administrator can not send reminders for "freeze after" questions, of which the deadline has passed.
- Lastly, we fixed some existing tests which failed due to our changes.

## How Has This Been Tested?

We have changed the existing tests according to changes in code. We want advice to create new tests.

## Screenshots:
Στιγμιότυπο οθόνης (312)
Στιγμιότυπο οθόνης (310)

## Checklist

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [ ] My change is listed in the CHANGELOG.rst if appropriate.

### CODE REVIEW

File: src/pretalx/common/mixins/forms.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
+import datetime
```
Comment (rixx) [inline src/pretalx/common/mixins/forms.py:1]: Could you please use `import datetime as dt` here? The whole codebase uses this pattern (due to the concerns outlined [here](https://adamj.eu/tech/2019/09/12/how-i-import-pythons-datetime-module/)).

File: src/pretalx/common/mixins/forms.py [hunk 2]
```diff
@@ -87,6 +89,34 @@ class QuestionFieldsMixin:
     def get_field(self, *, question, initial, initial_object, readonly):
         from pretalx.submission.models import QuestionVariant
 
+        tz = pytz.timezone(question.event.timezone)
```
Comment (rixx) [inline src/pretalx/common/mixins/forms.py:92]: You don't need to do this – timezone data in the database is localised, as is `now()`. You can compare the two directly without any other timezone considerations.

File: src/pretalx/common/mixins/forms.py [hunk 3]
```diff
@@ -87,6 +89,34 @@ class QuestionFieldsMixin:
...
+
+        # Check (with question_required field) if the question should be required
+        if question.question_required == "require":
```
Comment (rixx) [inline src/pretalx/common/mixins/forms.py:95]: Regarding this require/disable logic, I think this is not the best place to put it. What I'd like to see is a `required` property (or `cached_property`) on the `Question` class, so that we can keep using `question.required`. If that doesn't make sens to you, let me know and I can make the change myself.

Plus a similar property for `readonly`.

File: src/pretalx/orga/forms/cfp.py [hunk 4]
```diff
@@ -153,22 +156,45 @@ def __init__(self, *args, event=None, **kwargs):
         ):
             self.fields["is_public"].disabled = True
 
+    def clean(self):
+        deadline = self.cleaned_data["deadline"]
+        question_required = self.cleaned_data["question_required"]
+        if (deadline == "" or deadline is None) and (
+            question_required == "require after" or question_required == "freeze after"
+        ):
+            raise forms.ValidationError(
```
Comment (rixx) [inline src/pretalx/orga/forms/cfp.py:165]: Please make the errors translated strings using `_()`.

File: src/pretalx/orga/forms/cfp.py [hunk 5]
```diff
@@ -153,22 +156,45 @@ def __init__(self, *args, event=None, **kwargs):
         ):
             self.fields["is_public"].disabled = True
 
+    def clean(self):
+        deadline = self.cleaned_data["deadline"]
+        question_required = self.cleaned_data["question_required"]
+        if (deadline == "" or deadline is None) and (
```
Comment (rixx) [inline src/pretalx/orga/forms/cfp.py:162]: You can just use `if deadline` here.

File: src/pretalx/orga/forms/cfp.py [hunk 6]
```diff
@@ -153,22 +156,45 @@ def __init__(self, *args, event=None, **kwargs):
...
+            question_required == "require after" or question_required == "freeze after"
+        ):
+            raise forms.ValidationError(
+                "If you select 'freeze after deadline' or 'require after deadline' choice you "
+                + "should select the date and time deadline."
+            )
+        if (deadline != "" and deadline is not None) and (
```
Comment (rixx) [inline src/pretalx/orga/forms/cfp.py:169]: You can just use `if deadline` here.

File: src/pretalx/orga/forms/cfp.py [hunk 7]
```diff
@@ -153,22 +156,45 @@ def __init__(self, *args, event=None, **kwargs):
...
+            question_required == "none" or question_required == "require"
+        ):
+            raise forms.ValidationError(
```
Comment (rixx) [inline src/pretalx/orga/forms/cfp.py:172]: Please make the errors translated strings using `_()`.

File: src/pretalx/orga/forms/cfp.py [hunk 8]
```diff
@@ -279,7 +305,7 @@ def __init__(self, *args, instance, user, **kwargs):
                 _(
                     """Hi!
 
-This is an access code for the {event} CfP."""
+    This is an access code for the {event} CfP."""
```
Comment (rixx) [inline src/pretalx/orga/forms/cfp.py:308]: Accidentally added whitespace? Please remove again :)

File: src/pretalx/orga/forms/cfp.py [hunk 9]
```diff
@@ -421,7 +447,18 @@ class ReminderFilterForm(QuestionFilterForm):
...
+        questions = Question.objects.all()
+        now = datetime.datetime.now()
+        # We want to exclude questions with "freeze after", the deadlines of which have passed
+        questions_frozen = set()
+        for q in questions:
+            tz = pytz.timezone(q.event.timezone)
+            if q.question_required == "freeze after" and q.deadline < tz.localize(now):
+                questions_frozen.add(q.id)
+
+        questions.filter(id__in=questions_frozen).delete()
```
Comment (rixx) [inline src/pretalx/orga/forms/cfp.py:459]: You are **deleting** questions here, which surely can't be intentional!

File: src/pretalx/orga/forms/cfp.py [hunk 10]
```diff
@@ -421,7 +447,18 @@ class ReminderFilterForm(QuestionFilterForm):
     )
 
     def get_question_queryset(self):
-        return Question.objects.filter(
+        questions = Question.objects.all()
+        now = datetime.datetime.now()
+        # We want to exclude questions with "freeze after", the deadlines of which have passed
+        questions_frozen = set()
```
Comment (rixx) [inline src/pretalx/orga/forms/cfp.py:453]: This should be doable in a single query, without looping, I think. Something like this:

```python
Question.objects.filter(
    event=self.event,
    target__in=["speaker", "submission"],
).exclude(Q(question_required="freeze after") & Q(deadline__lt=now()))
```

File: src/pretalx/orga/templates/orga/cfp/question_form.html [hunk 11]
```diff
@@ -7,11 +7,17 @@
...
+        <script defer src="{% static "cfp/js/animateFormset.js" %}"></script>
+        <script defer src="{% static "vendored/moment-with-locales.js" %}"></script>
+        <script defer src="{% static "vendored/moment-timezone-with-data-2012-2022.js" %}"></script>
+        <script defer src="{% static "vendored/datetimepicker/bootstrap-datetimepicker.js" %}"></script>
+        <script defer src="{% static "orga/js/datetimepicker.js" %}"></script>
+        <script defer src="{% static "orga/js/questionRequiredField.js" %}"></script>
     {% endcompress %}
     <h2>
-        {% if form.instance.question %}{% translate "Question" %}: {{ form.instance.question }}{% else %}{% translate "New question" %}{% endif %}
+        {% if form.instance.question %}{% translate "Question" %}: {{ form.instance.question }}{% else %}
```
Comment (rixx) [inline src/pretalx/orga/templates/orga/cfp/question_form.html:19]: Please revert this change.

File: src/pretalx/orga/templates/orga/cfp/question_form.html [hunk 12]
```diff
@@ -27,21 +33,23 @@ <h2>
         {% endif %}
 
         {% if action != 'create' %}
-          {% if question.active %}
-            <div class="alert alert-info col-md-9 offset-md-3">
+            {% if question.active %}
```
Comment (rixx) [inline src/pretalx/orga/templates/orga/cfp/question_form.html:36]: Could you please revert indentation changes like this that do not add to the PR? It makes it harder to see differences and makes it harder to use `git blame` in the future.

File: src/pretalx/orga/templates/orga/cfp/question_view.html [hunk 13]
```diff
@@ -28,7 +29,7 @@
             <tr>
                 <th>{% translate "Question" %}</th>
                 <th>{% translate "Target" %}</th>
-                <th>{% translate "required" %}</th>
+                <th>{% translate "require" %}</th>
```
Comment (rixx) [inline src/pretalx/orga/templates/orga/cfp/question_view.html:32]: Is this a necessary change? It forces all translations to update, so I'd like to be conservative if not.

File: src/pretalx/orga/templates/orga/cfp/question_view.html [hunk 14]
```diff
@@ -45,13 +46,21 @@
...
+                        {% if question.question_required == "require" %}
+                            {% translate "always" %}
+                        {% elif question.question_required == "require after"%}
+                            {% translate "after" %}
+                        {% elif question.question_required == "freeze after"%}
+                            {% translate "before" %}
+                        {% else %}
+                            {% translate "never" %}
+                        {% endif %}
+                        {% if question.deadline %}
```
Comment (rixx) [inline src/pretalx/orga/templates/orga/cfp/question_view.html:58]: Alternative suggestion: Let's keep the "required" column to a yes/no indicator that reflects if the question is currently required (using the `required` property I suggested earlier). Then add another indicator, a clock or the like, to signal if this property is based on a deadline or not.

Otherwise, this part of the table will start being unreasonably wide. We could put the deadline in the alt text for the clock icon.

File: src/pretalx/orga/templates/orga/cfp/question_view.html [hunk 15]
```diff
@@ -45,13 +46,21 @@
...
+                            {{ question.deadline|timezone:question.event.timezone }}
```
Comment (rixx) [inline src/pretalx/orga/templates/orga/cfp/question_view.html:59]: No need to add timezones here, all times are already localised.

File: src/pretalx/submission/migrations/0058_remove_question_required.py [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("submission", "0057_auto_20210508_1308"),
+    ]
+
+    operations = [
+        migrations.RemoveField(
```
Comment (rixx) [inline src/pretalx/submission/migrations/0058_remove_question_required.py:13]: If we do this, we would lose all data on existing questions in pretalx – we can't do this, it would break events that are running right now. There needs to be a data migration in between the two migrations that transfers the `required` state to the new `question_required` field.

If you need help with this, please let me know, and/or just drop the migration files from this PR, and I'll take care of it.

File: src/pretalx/submission/models/question.py [hunk 17]
```diff
@@ -61,6 +61,20 @@ class QuestionTarget(Choices):
     ]
 
 
+class QuestionRequired(Choices):
+    NONE = "none"
```
Comment (rixx) [inline src/pretalx/submission/models/question.py:65]: Please always use these attributes for comparison in the rest of the code base. Comparing to `"require after"` is fragile when it would be better to compare to `QuestionRequired.REQUIRE_AFTER`.

File: src/pretalx/submission/models/question.py [hunk 18]
```diff
@@ -97,6 +119,32 @@ class Question(LogMixin, models.Model):
             "Do you require an answer from every speaker or for every session?"
         ),
     )
+
+    deadline = models.DateTimeField(
+        null=True,
+        blank=True,
+        verbose_name=_("deadline"),
+        help_text=_(
+            "This field is required for 'require after' and 'freeze after' options "
```
Comment (rixx) [inline src/pretalx/submission/models/question.py:128]: More user friendly suggestion (users don't care what a field is etc): "Set a deadline to make this question required or to freeze changes after the given date."

File: src/tests/api/test_api_questions.py [hunk 19]
```diff
@@ -29,7 +29,8 @@ def test_question_serializer(answer):
         "id",
         "variant",
         "question",
-        "required",
+        "question_required",
```
Comment (rixx) [inline src/tests/api/test_api_questions.py:32]: Could you add these changes to the API documentation, please?

**Review by rixx — CHANGES_REQUESTED** [review]
Thank you for the PR! It's generally looking good, and I left some review comments. Generally I'm happy to merge this as soon as you adjust the PR to address the review comments – it's up to you how much of the more complex problems you want to tackle. I'm happy to take care of anything that is too much for you at this time – especially as I realise that it's due to me being slow that this PR is still open ;)

One major concern that occurred to me is that with this new mode, it's impossible to freeze a question that has been made mandatory after a certain date. That seems unfortunate, and like we are conflating two different concepts. I think a good solution would be to remove the "freeze after" option and move it to a different new `DateTimeField`.

Since that is a bit of work and nothing that was suggested ahead of time, please let me know if you have any interest in this. The easiest way forward would be to remove the "freeze" functionality from this PR, and to take care of the problem later on.

File: src/pretalx/common/mixins/forms.py (hunk 2 above)
Comment (AthinaDavari) [reply src/pretalx/common/mixins/forms.py:92]: I changed this:
```
tz = pytz.timezone(question.event.timezone)
now = datetime.datetime.now()
if question.deadline > tz.localize(now):
                require_question = False
```
To this:
```
from django.utils import timezone
now = timezone.now()
if question.deadline > now:
                require_question = False
```
Is this what you wanted?
When i didn't use ```pytz.timezone()``` and ```tz.localize()``` in the comparison,
it thew an error: ```can't compare offset-naive and offset-aware datetimes```

File: src/pretalx/common/mixins/forms.py (hunk 1 above)
Comment (AthinaDavari) [reply src/pretalx/common/mixins/forms.py:1]: I fixed the files ```src/pretalx/common/mixins/forms.py``` and ```src/pretalx/orga/forms/cfp.py```. Now the files have the correct import pattern.

File: src/pretalx/orga/forms/cfp.py (hunk 4 above)
Comment (AthinaDavari) [reply src/pretalx/orga/forms/cfp.py:165]: 🗺I fixed that for the two ValidationError messages in the file ```src/pretalx/orga/forms/cfp.py```.🗺

File: src/pretalx/common/mixins/forms.py (hunk 2 above)
Comment (rixx) [reply src/pretalx/common/mixins/forms.py]: Yes, that is what I meant. How did you create the data that resulted in this error? Usually, all datetime data that goes through Django is localised (see also: the cfp deadline), and there should never be naive datetimes in our database.

File: src/pretalx/orga/forms/cfp.py (hunk 5 above)
Comment (AthinaDavari) [reply src/pretalx/orga/forms/cfp.py:162]: I made simpler the if conditions using ```not deadline``` and ```deadline```.

File: src/pretalx/common/mixins/forms.py (hunk 2 above)
Comment (AthinaDavari) [reply src/pretalx/common/mixins/forms.py]: With the changes i made and i add in the previous message, it works. 
When i just removed ```pytz.timezone()``` and ```tz.localize()``` it didn't work.

File: src/pretalx/submission/models/question.py (hunk 18 above)
Comment (AthinaDavari) [reply src/pretalx/submission/models/question.py]: I replased the message. 🎀

File: src/pretalx/orga/forms/cfp.py (hunk 6 above)
Comment (AthinaDavari) [reply src/pretalx/orga/forms/cfp.py]: I made simpler the if condition using ```deadline```.

File: src/pretalx/orga/forms/cfp.py (hunk 7 above)
Comment (AthinaDavari) [reply src/pretalx/orga/forms/cfp.py:172]: 🗺I fixed that for the ValidationError message in the file src/pretalx/orga/forms/cfp.py.🗺

File: src/pretalx/orga/forms/cfp.py (hunk 8 above)
Comment (AthinaDavari) [reply src/pretalx/orga/forms/cfp.py]: Format fixed 🎨

File: src/pretalx/orga/templates/orga/cfp/question_view.html (hunk 15 above)
Comment (AthinaDavari) [reply src/pretalx/orga/templates/orga/cfp/question_view.html]: Fixed🧹🔥

File: src/pretalx/orga/templates/orga/cfp/question_view.html (hunk 13 above)
Comment (AthinaDavari) [reply src/pretalx/orga/templates/orga/cfp/question_view.html:32]: Fixed.

File: src/pretalx/orga/forms/cfp.py (hunk 9 above)
Comment (nataliaKat) [reply src/pretalx/orga/forms/cfp.py]: Yes, I'm sorry, my bad.

File: src/pretalx/orga/templates/orga/cfp/question_form.html (hunk 12 above)
Comment (AthinaDavari) [reply src/pretalx/orga/templates/orga/cfp/question_form.html:36]: Hi, we tried to fix it, but we did not find a way to do it. 
We found some possible ways to recover an old version of a file on the Internet and tested them in a test repository. Some of them worked, some did not.
But we did not find a way to show the original author of the line with the git blame. 
Do you have a suggestion on how to do this?

**Comment by nataliaKat** [comment]
You are right, freeze after is irrelevant with required. We can add the extra freeze_after datetime field which will only disable the question (and make it optional) and remove the freeze_after option.
We hope we will make it till the end of the week since we are a bit more familiar with the project now.

File: src/pretalx/orga/templates/orga/cfp/question_form.html (hunk 12 above)
Comment (rixx) [reply src/pretalx/orga/templates/orga/cfp/question_form.html:36]: Don't worry about it – I'll squash/rebase some of the commits so that it will work out.

**Comment by rixx** [comment]
Sounds good! Let's agree on this: take another day (Wednesday) to make the changes you have the time for, and after Wednesday I'll merge the PR and make the necessary changes (if any remain) afterwards. So you'll be guaranteed your merged PR by Friday :)

File: src/pretalx/common/mixins/forms.py (hunk 3 above)
Comment (AthinaDavari) [reply src/pretalx/common/mixins/forms.py]: ✨  I added ```required``` property to ```Question``` class.

File: src/pretalx/orga/templates/orga/cfp/question_view.html (hunk 14 above)
Comment (AthinaDavari) [reply src/pretalx/orga/templates/orga/cfp/question_view.html]: - [x] ```required``` table column like before changes
- [X] ```deadline``` table column

File: src/pretalx/submission/migrations/0058_remove_question_required.py (hunk 16 above)
Comment (nataliaKat) [reply src/pretalx/submission/migrations/0058_remove_question_required.py:13]: We are not sure onhow to make the migration so we will just delete the files and we won't commit the last ones with the ```freeze_after``` field.

File: src/pretalx/orga/templates/orga/cfp/question_view.html (hunk 14 above)
Comment (AthinaDavari) [reply src/pretalx/orga/templates/orga/cfp/question_view.html]: Στιγμιότυπο οθόνης (315)

Comment (AthinaDavari) [reply src/pretalx/orga/templates/orga/cfp/question_view.html]: Στιγμιότυπο οθόνης (316)

File: src/pretalx/submission/models/question.py (hunk 17 above)
Comment (AthinaDavari) [reply src/pretalx/submission/models/question.py:66]: Fixed.

File: src/pretalx/orga/templates/orga/cfp/question_form.html (hunk 11 above)
Comment (AthinaDavari) [reply src/pretalx/orga/templates/orga/cfp/question_form.html:19]: Here we have its same with the next comment.
We did not find a way to show the original author of the line with the git blame.

Comment (rixx) [reply src/pretalx/orga/templates/orga/cfp/question_form.html:19]: As I said: Don't worry about it. I'll squash the changes into a single or a few commits before merging, and after that, git blame will work as expected, as if your change had never happened.

File: src/tests/api/test_api_questions.py (hunk 19 above)
Comment (nataliaKat) [reply src/tests/api/test_api_questions.py:32]: API documentation is ready.

**Comment by nataliaKat** [comment]
Hello, we've finished with the changes! We managed to make ```freeze_after``` a separate field. We really appreciated your help and quick responses, thank you very much!!! If there is anything else that needs to be done, please let us know. One last question: are tests ok? Should we add any new tests, since we've added new functionality? If yes, could you give us some guidelines?

**Comment by rixx** [comment]
Alright, I finished work for today, now I'll look at this. In case you're interested in what I'm doing with your PR, I'm going to document all my steps here. But first off: Thank you so much! I really appreciate it, and any modifications are not criticism, just making sure everything is in place.

So first off, I fixed the master branch's failing tests and rebased your changes onto the master branch (that was the push from 052741d to faff0fc).

**Comment by rixx** [comment]
I've put my steps so far in the issue comment above (edited for each step, in case you missed it). I'll take a break for about two hours now, and will then finish testing the change, then write the migrations, and will then merge the PR.

**Comment by rixx** [comment]
And with [these migrations](https://github.com/pretalx/pretalx/pull/1123/commits/bbbcab0be7116cafd8b5fc8fa10e0f7b601978dc), we're done! Ready to merge, and you're official contributors. Thank you again!

Do you want to pick up any other issues? I can go look at the issue tracker for suggestions, if you're interested.

**Review by rixx — APPROVED** [review]

**Comment by nataliaKat** [comment]
Thanks again for all this help and patience, it was really valuable for us, since it was our first OS contribution!!! If there is any way we can help with test coverage please let us know. Our spring exams are about to start, so we will be a bit busy for July, however we hope to find time during the next months and pick some issues!
